transactions/transactions_service.go

```diff
@@ -0,0 +1,690 @@
+			return errors.New("failed to cast event")
+		}
+
+		err := svc.db.Transaction(func(tx *gorm.DB) error {
+			var dbTransaction db.Transaction
+
+			result := tx.Find(&dbTransaction, &db.Transaction{
+				Type:        constants.TRANSACTION_TYPE_INCOMING,
+				PaymentHash: lnClientTransaction.PaymentHash,
+			})
+
+			if result.RowsAffected == 0 {
+				// Note: brand new payments (keysend only) cannot be associated with an app
+				var metadata string
+				if lnClientTransaction.Metadata != nil {
+					metadataBytes, err := json.Marshal(lnClientTransaction.Metadata)
+					if err != nil {
+						logger.Logger.WithError(err).Error("Failed to serialize transaction metadata")
+						return err
+					}
+					metadata = string(metadataBytes)
+				}
+				var expiresAt *time.Time
+				if lnClientTransaction.ExpiresAt != nil {
+					expiresAtValue := time.Unix(*lnClientTransaction.ExpiresAt, 0)
+					expiresAt = &expiresAtValue
+				}
+				dbTransaction = db.Transaction{
+					Type:            constants.TRANSACTION_TYPE_INCOMING,
+					AmountMsat:      uint64(lnClientTransaction.Amount),
+					PaymentRequest:  lnClientTransaction.Invoice,
+					PaymentHash:     lnClientTransaction.PaymentHash,
+					Description:     lnClientTransaction.Description,
+					DescriptionHash: lnClientTransaction.DescriptionHash,
+					ExpiresAt:       expiresAt,
+					Metadata:        metadata,
+				}
+				err := tx.Create(dbTransaction).Error
+				if err != nil {
+					logger.Logger.WithFields(logrus.Fields{
+						"payment_hash": lnClientTransaction.PaymentHash,
+					}).WithError(err).Error("Failed to create transaction")
+					return err
+				}
+			}
+
+			settledAt := time.Now()
+			fee := uint64(lnClientTransaction.FeesPaid)
+
+			err := tx.Model(&dbTransaction).Updates(&db.Transaction{
+				FeeMsat:   &fee,
+				Preimage:  &lnClientTransaction.Preimage,
+				State:     constants.TRANSACTION_STATE_SETTLED,
+				SettledAt: &settledAt,
+			}).Error
+			if err != nil {
+				logger.Logger.WithFields(logrus.Fields{
+					"payment_hash": lnClientTransaction.PaymentHash,
+				}).WithError(err).Error("Failed to update transaction")
+				return err
+			}
+
+			logger.Logger.WithField("id", dbTransaction.ID).Info("Marked incoming transaction as settled")
+			return nil
+		})
+
+		if err != nil {
+			logger.Logger.WithFields(logrus.Fields{
+				"payment_hash": lnClientTransaction.PaymentHash,
+			}).WithError(err).Error("Failed to execute DB transaction")
+			return err
+		}
+	case "nwc_payment_sent":
+		lnClientTransaction, ok := event.Properties.(*lnclient.Transaction)
+		if !ok {
+			logger.Logger.WithField("event", event).Error("Failed to cast event")
+			return errors.New("failed to cast event")
+		}
+
+		var dbTransaction db.Transaction
+		result := svc.db.Find(&dbTransaction, &db.Transaction{
+			Type:        constants.TRANSACTION_TYPE_OUTGOING,
+			PaymentHash: lnClientTransaction.PaymentHash,
+		})
+
+		if result.RowsAffected == 0 {
```

Comment:
What if an LND user did a payment externally? Then #281 would send a `nwc_payment_sent ` event to return an error here

Reply:
fixed